src/workerd/api/sockets.c++

```diff
@@ -263,14 +263,30 @@ jsg::Ref<Socket> connectImpl(
 }
 
 jsg::Promise<void> Socket::close(jsg::Lock& js) {
-  // Forcibly close the readable/writable streams.
-  auto cancelPromise = readable->getController().cancel(js, kj::none);
-  auto abortPromise = writable->getController().abort(js, kj::none);
-  // The below is effectively `Promise.all(cancelPromise, abortPromise)`
-  return cancelPromise.then(js, [abortPromise = kj::mv(abortPromise), this](jsg::Lock& js) mutable {
-    return abortPromise.then(js, [this](jsg::Lock& js) {
-      resolveFulfiller(js, kj::none);
-      return js.resolvedPromise();
+  if (isClosing) {
+    return closedPromiseCopy.whenResolved(js);
+  }
+
+  isClosing = true;
+
+  auto flushPromise = js.resolvedPromise();
+  if (!writable->getController().isClosedOrClosing()) {
+    flushPromise = writable->getController().flush(js);
+  }
+
+  return flushPromise.then(js, [this](jsg::Lock& js) {
+    // Forcibly abort the readable/writable streams.
+    auto cancelPromise = readable->getController().cancel(js, kj::none);
+    auto abortPromise = writable->getController().abort(js, kj::none);
+    // The below is effectively `Promise.all(cancelPromise, abortPromise)`
+    return cancelPromise.then(js, [abortPromise = kj::mv(abortPromise), this](jsg::Lock& js) mutable {
+      return abortPromise.then(js, [this](jsg::Lock& js) {
+        // Wait until the socket connects (successfully or otherwise)
+        return openedPromiseCopy.then(js, [this](jsg::Lock& js) {
+          resolveFulfiller(js, kj::none);
+          return js.resolvedPromise();
+        }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
+      }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
     }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
   }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
```

Comment:
... So... much... nesting... ah! Non-blocking but *might* we be able to simplify this a bit with a little more chaining instead of nesting?

Reply:
Ahh yeah, I was going to fix that but forgot. Will do before merging.

Reply:
Actually now that I look at it I'm not sure how to do that. How do you chain jsg::Promises? I thought I would be able to use coroutines but seems they are not supported for jsg::Promises yet.

Reply:
I think he means that instead of

`promise1.then([]() { return promise2.then(...); })`

you can do

`.promise1.then([]() { return promise2; }).then(...)`

This avoids so much nesting.

On a related note, you don't need every `.then()` to have an error handler, you only need to add a `.catch_()` at the end of the chain...